### ISSUE TITLE
Volume claim remains pending forever VMFS

### ISSUE THREAD

**Issue by macdja38** [rancher/rancher#14724]
**Rancher versions:**
rancher/server or rancher/rancher: rancher/rancher image is sha256:975f0d475e475d86e5e20c75ad37630407236066953150caafc5677985da88a8
rancher/agent or rancher/rancher-agent: v2.0.4
This is what you see when you click the version in the bottom left of the rancher ui.
Rancher | v2.0.4
User Interface | v2.0.53
Helm | dev
Machine | dev

**Infrastructure Stack versions:**
healthcheck:
ipsec:
network-services:
scheduler:
kubernetes (if applicable):
Client Version: version.Info{Major:"1", Minor:"9", GitVersion:"v1.9.8", GitCommit:"c138b85178156011dc934c2c9f4837476876fb07", GitTreeState:"clean", BuildDate:"2018-05-21T19:01:12Z", GoVersion:"go1.9.3", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"10", GitVersion:"v1.10.3", GitCommit:"2bba0127d85d5a46ab4b778548be28623b32d0b0", GitTreeState:"clean", BuildDate:"2018-05-21T09:05:37Z", GoVersion:"go1.9.3", Compiler:"gc", Platform:"linux/amd64"}

**Docker version: (`docker version`,`docker info` preferred)**
Note that there may be typos.
Client:
version 17.03.2-ce
api version: 1.27
Go version: go1.7.5
git commit: f5ec1e2
built: tue jan 27th 2017

server
version: 17.03.2-ce
api version: 1.27
go version: go1.7.5
git commit f5ec1e2
built tue jay 27th 2017

**Operating system and kernel: (`cat /etc/os-release`, `uname -r` preferred)**
rancheros version v1.4.0, kernel 4.14.32-rancher2

**Type/provider of hosts: (VirtualBox/Bare-metal/AWS/GCE/DO)**
vSphere local backed by VMFS file system.

**Setup details: (single node rancher vs. HA rancher, internal DB vs. external DB)**
rancher on one ubuntu vm with the single image docker install, vSphere cluster setup on one machine with 3 etc/control nodes and 5 workers.

**Environment Template: (Cattle/Kubernetes/Swarm/Mesos)**
Not sure how to check.

**Steps to Reproduce:**
1. Create a storage class manually through the CLI
```
apiVersion: storage.k8s.io/v1
kind: StorageClass
metadata:
  creationTimestamp: 2018-07-23T17:35:55Z
  name: thin-1-host
  resourceVersion: "5388433"
  selfLink: /apis/storage.k8s.io/v1/storageclasses/thin-1-host
  uid: da7d6727-eo83-11e8-eoau-000c290e8b79
parameters:
  datastore: D1
  diskformat: thin
  hostFailuresToTolerate: "1"
provisioner: kubernetes.io/vsphere-volume
reclaimPolicy: Delete
volumeBindingMode: Immediate
```

2. Create a volume through the ui within a project namespace using add volume
put the settings as
Name: whatever
Namespace: whatever
Storage Class: thin-1-host
Capacity: 30GiB (can be whatever)

3. Click create

**Results:**
 It will stay pending forever

If I'm missing any necessary data please let me know.

**Comment by ntwrkguru** [comment]
Ditto. Rancher 2.0.6 on a single CentOS 7 VM against vSphere 6.5.


Rancher | v2.0.6
-- | --
User Interface | v2.0.61
Helm | v2.8.0-rancher3
Machine | v0.14.0-rancher4-1

**Comment by aemneina** [comment]
Can you check the comment here to see if it resolves your issue:
https://github.com/rancher/rancher/issues/14334#issuecomment-401770645

**Comment by macdja38** [comment]
I have tried that to no avail.

**Comment by MaxDiOrio** [comment]
Same here on rancher 2.0.6.  vSphere Cloud Provider configured on the cluster.  

Create a volume at the cluster level and it immediately comes up at Available, but never created the VMDK in vSphere.  

When I go to create a claim against the existing persistent volume, it stays pending.

**Comment by lynic** [comment]
I have the same issue here with 2.0.6 using rancher-vmware.iso .

**Comment by lynic** [comment]
Well, after reading #13115 , it works for me now.

**Comment by ntwrkguru** [comment]
@lynic Do you mean it works out-of-the-box now? If so, what build? Because if "it works for me now" is only after you use some custom cloud provider, then that's still not a fix in my opinion.

**Comment by lynic** [comment]
I use the custom cloud provider, and put the configuration to yaml.

**Comment by ntwrkguru** [comment]
I refuse to accept that as a solution. This is vSphere, not some obscure hypervisor/cloud setup.

**Comment by macdja38** [comment]
What did you use as the custom cloud provider if I may ask @lynic ?

**Comment by MaxDiOrio** [comment]
This seems to be the case for all cloud providers. Vmware works perfectly fine without the cloud provider as long as you don't want persistent storage. It's the persistent storage piece is why I set up the cloud provider and it does absolutely nothing to actually provision the volumes in vSphere

Sent from Nine<http://www.9folders.com/>
________________________________

**Comment by MaxDiOrio** [comment]
@lynic - do you mean you can provision persistent volume now?  Still can't here.  What did you use for your cloud provider info?

**Comment by lynic** [comment]
The same as comment from sangeethah in #13115

**Comment by macdja38** [comment]
@lynic so you didn't use the built in v-sphere config you wrote your own, and with the write your own thing you were able to provision disks on a VMFS file system?

**Comment by MaxDiOrio** [comment]
There is no built in vSphere cloud provider config.

**Comment by macdja38** [comment]
@MaxDiOrio Sorry, I mean the vSphere node driver.

**Comment by MaxDiOrio** [comment]
@lync.  Thank you!!!!   There were two settings I needed to change, insecure-flag kept getting set to false, even though my initial config was set to "1" for true.  And I added the username/password under virtual_center, as well as under global.  virtual_center is supposed to inherit from global, but I'm guessing it doesn't?

**Comment by MaxDiOrio** [comment]
@macdja38 - I bet he's using the built in vSphere node driver - I was too but the vsphereCloudProvider needed to be in a specific format that seems different.  Plus in once of their documents, it had a typo in it.  After changing my config to the one @lync pointed to, I can now provision persistent storage and Rancher created the VMDKs finally.

**Comment by macdja38** [comment]
@MaxDiOrio I'm confused about the difference between the vSphere node driver and vsphereCloudProvider.

If I have it setup using the node driver on a local vSphere install what exactly do I have to do to get this working?

I think the typo should be solved by manually creating the storage class through the CLI, but I'm not sure.

Thanks for your time.

**Comment by macdja38** [comment]
@MaxDiOrio thank you very much for the detailed instructions, I'll give that a try tomorrow!

@sangeethah apologies for the disruption, you seem to have assigned yourself to the original issue that added this node provider, is there any chance you could look into modifying it so something like the above is not necessary in the future?

**Comment by superseb** [comment]
Configuring cloud providers (and in this case vSphere) is documented here: https://rancher.com/docs/rke/v0.1.x/en/config-options/cloud-providers/vsphere/

What exactly is missing from these instructions?

**Comment by ntwrkguru** [comment]
@superseb, for starters, the fact that this is the first time I've seen those instructions and the fact that the casual user wouldn't know to go to rke to look for something that *appears* to work directly from within the Rancher UI. Put some docs ON THE PAGE, or do something else to let the user know that, while it looks like it works, the UI doesn't actually do anything if that bit is missing. Ideally, one wouldn't need to augment the wonderful UI with a bunch of custom YAML, but that's probably more of an enhancment.

**Comment by char-iot** [comment]
@superseb doc works fine now after v2.0.6.

This is about after you created the storage class. When you try to create a volume, you'll get error in vCenter, it keeps trying to create the same volume for many times even though first time was successful.

**Comment by superseb** [comment]
@ntwrkguru Not sure why the tone, but we are all people here trying to help.

@MaxDiOrio I asked my colleague who has created the docs to take a look at your comments.

And we will be listing these instructions on https://rancher.com/docs/rancher/v2.x/en/cluster-provisioning/rke-clusters/options/cloud-providers/ as well, just need to figure out what should go where exactly.

**Comment by ntwrkguru** [comment]
Didn’t meant for it to come across as a tone. Sorry. My points stand however. The entire reason this is an issue was that it seems to work “out of the box”, but actually requires a lot of manual intervention which, until recently, was poorly documented, if at all.

Sent from my iPhone

On Aug 14, 2018, at 3:58 PM, Sebastiaan van Steenis <notifications@github.com<mailto:notifications@github.com>> wrote:


@ntwrkguru<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_ntwrkguru&d=DwMFaQ&c=HAkYuh63rsuhr6Scbfh0UjBXeMK-ndb3voDTXcWzoCI&r=y8f5-ZKbifhgVsSVihKTuG2bbsBaiqELScS4YCpX8v8&m=0LAfL1Pq5GZOJ8TGcZGhtLmetYrRmbjLD5sfMfaOur4&s=YSbtpuaW5MmmBZdr8R8lUwH2JLldQ-PaIqM87t_ZYug&e=> Not sure why the tone, but we are all people here trying to help.

@MaxDiOrio<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_MaxDiOrio&d=DwMFaQ&c=HAkYuh63rsuhr6Scbfh0UjBXeMK-ndb3voDTXcWzoCI&r=y8f5-ZKbifhgVsSVihKTuG2bbsBaiqELScS4YCpX8v8&m=0LAfL1Pq5GZOJ8TGcZGhtLmetYrRmbjLD5sfMfaOur4&s=xQn3FTSI_PZLEEkW_5SSvhgHyH4F0XRnAfrhuPj9zcQ&e=> I asked my colleague who has created the docs to take a look at your comments.

And we will be listing these instructions on https://rancher.com/docs/rancher/v2.x/en/cluster-provisioning/rke-clusters/options/cloud-providers/<https://urldefense.proofpoint.com/v2/url?u=https-3A__rancher.com_docs_rancher_v2.x_en_cluster-2Dprovisioning_rke-2Dclusters_options_cloud-2Dproviders_&d=DwMFaQ&c=HAkYuh63rsuhr6Scbfh0UjBXeMK-ndb3voDTXcWzoCI&r=y8f5-ZKbifhgVsSVihKTuG2bbsBaiqELScS4YCpX8v8&m=0LAfL1Pq5GZOJ8TGcZGhtLmetYrRmbjLD5sfMfaOur4&s=3-vUqHjkUbwIIDqe7OxC3e4vvic95nk_qlBvbhJfIek&e=> as well, just need to figure out what should go where exactly.

**Comment by macdja38** [comment]
@superseb Hey, have not gotten it working with those instructions yet.

I'm decently new to vSphere and vCenter and all that so some of the basic things take me a while to find. One thing that confused me for example is the casing of disk.enableUUID = true, I found online that it seems the casing of true doesn't matter, but I'm still not sure if EnableUUID or enableUUID is correct, the docs seem to have both? https://rancher.com/docs/rke/v0.1.x/en/config-options/cloud-providers/vsphere/#enabling-disk-uuids-for-vsphere-vms

Basically I never even found the RKE documentation, I was just very confused about why the Rancher 2.0 documentation was so lacking when I was searching through it. I'll give it a shot now, but I feel these things really should be in a UI, and there should be an error if you try and provision a storage thing without this setup, instead of stuff just hanging forever.

**Comment by janeczku** [comment]
@MaxDiOrio 

> There's a typo on the page you referenced. In the example code, "datacenters" should really be "datacenter"

`datacenters` is the *new* `datacenter` 😉 the latter is [deprecated](https://github.com/kubernetes/kubernetes/blob/release-1.10/pkg/cloudprovider/providers/vsphere/vsphere.go#L112-L115).

> When I used insecure-flag: "1" which a lot of documentation lists, it was converted to insecure-flag: "false". 

These documentations refer to the format of the ["INI-style"](https://godoc.org/gopkg.in/gcfg.v1) `vsphere.conf` file, which is much more loose in terms of what gets interpreted as boolean value ("true", "yes", "on" or "1" - everything goes). As documented [here](https://rancher.com/docs/rke/v0.1.x/en/config-options/cloud-providers/vsphere/#global) the `insecure-flag` in the cluster YAML-file's vSphere cloud provider directive is a boolean field, which in the YAML scheme of things translates to: must specify scalars of either `true` or `false`.

**Comment by janeczku** [comment]
@ntwrkguru @macdja38 Imho the latest updates to our docs [here](https://rancher.com/docs/rke/v0.1.x/en/config-options/cloud-providers/vsphere) should guide you to a successful configuration of the cloud provider. If you think there are any gaps documentation wise, feel free to point them out and we will address. I also think you have a valid point about a tighter UI integration and we will discuss internally to see if we can make the vSphere cloud provider configuration a first class citizen in the UI.

**Comment by ntwrkguru** [comment]
Thank you @janeczku. I think part of the problem is the lack of any basic sanity checking done within kubernetes for storage classes. The UI output reflects the same status that the kubectl output does, which is pretty meaningless. Rancher is a victim there, but when folks are used to dealing with a UI for 99% of their tasks (i.e. many vSphere users), then not having full UI functionality is frustrating. When the UI appears to show success, but doesn't actually equal success, it's mildly infuriating.

**Comment by MaxDiOrio** [comment]
I agree. And the issues keep coming. So now that the volumes are provisioning, they aren't getting formatted and attached to the pod. The volume is getting added to the worker though.

VMware as a whole feels like a second class, thoroughly untested citizen.  Such a popular platform should work fairly seamlessly out of the box.

**Comment by superseb** [comment]
Regarding the custom cloud provider, this is what is shown when configuring it and can be followed to the documentation:

<img width="1371" alt="screen shot 2018-08-15 at 13 38 34" src="https://user-images.githubusercontent.com/2620179/44146594-9cb3ef90-a090-11e8-90b9-06b9e2fedcc6.png">

Regarding the current issue, please provide logging from `kubelet` and `kube-controller-manager` to identify what is going wrong with a volume.

**Comment by janeczku** [comment]
With regards to the discussion about better integration in the UI, i have created this ticket https://github.com/rancher/rancher/issues/15089

**Comment by deniseschannon** [comment]
I added links on the Rancher cloud provider docs to point to the RKE docs on how to configure for openstack/vsphere.

I'm going to close this issue as originally, there was no easy to configure custom cloud providers and the request for adding the config option in the UI will be tracked in #15089.
